src/Draco.Compiler/Internal/Semantics/Types/TypeInference.cs

```diff
@@ -0,0 +1,364 @@
+            var resultType = this.solver.Same(leftType, rightType).Result;
+            this.expressions[node] = resultType;
+        }
+
+        return this.Default;
+    }
+
+    public override Unit VisitIfExpr(ParseTree.Expr.If node)
+    {
+        // Inference in children
+        base.VisitIfExpr(node);
+
+        // Check if condition is bool
+        var conditionType = TypeChecker.TypeOf(this.db, node.Condition.Value);
+        this.solver.Same(conditionType, Type.Bool).ConfigureDiagnostic(diag => diag
+            .WithLocation(new Location.TreeReference(node.Condition.Value)));
+
+        var resultType = null as Type;
+        var leftType = TypeChecker.TypeOf(this.db, node.Then);
+        if (node.Else is null)
+        {
+            // Then part has to be of unit
+            this.solver.Same(leftType, Type.Unit)
+                .ConfigureDiagnostic(diag => diag
+                    .WithLocation(ExtractReturnLocation(node.Then))
+                    .AddRelatedInformation("an if-expression without an else branch must result in unit"));
+            resultType = leftType;
+        }
+        else
+        {
+            var rightType = TypeChecker.TypeOf(this.db, node.Else.Expression);
+            // TODO: Wrong constraint, we need "Common ancestor"
+            resultType = this.solver.Same(leftType, rightType)
+                .ConfigureDiagnostic(diags => diags
+                    .WithLocation(ExtractReturnLocation(node.Else.Expression))
+                    .AddRelatedInformation(
+                        format: "the other branch was inferred to be {0} here",
+                        formatArgs: new[] { leftType },
+                        location: ExtractReturnLocation(node.Then)))
+                .Result;
+        }
+        this.expressions[node] = resultType;
+
+        return this.Default;
+    }
+
+    public override Unit VisitWhileExpr(ParseTree.Expr.While node)
+    {
+        // Inference in children
+        base.VisitWhileExpr(node);
+
+        var conditionType = TypeChecker.TypeOf(this.db, node.Condition.Value);
+        this.solver.Same(conditionType, Type.Bool).ConfigureDiagnostic(diag => diag
+            .WithLocation(new Location.TreeReference(node.Condition.Value)));
+
+        return this.Default;
+    }
+
+    public override Unit VisitCallExpr(ParseTree.Expr.Call node)
+    {
+        // Inference in children
+        base.VisitCallExpr(node);
+
+        var calledType = TypeChecker.TypeOf(this.db, node.Called);
+        var argsType = node.Args.Value.Elements
+            .Select(a => TypeChecker.TypeOf(this.db, a.Value))
+            .ToImmutableArray();
+        // TODO: We want a "callable" constraint here, but for now we'll go around a bit
+        var clientType = new Type.Function(argsType, new Type.Variable(null));
+        var returnType = ((Type.Function)this.solver.Same(calledType, clientType).Result).Return;
+        this.expressions[node] = returnType;
```

Comment:
Missing positional info. Maybe add a `Callable` constraint, even if it just uses `Same` underneath, just so the call-site is nicer?